### ISSUE TITLE
Don't store state data in $XDG_CONFIG_HOME

### ISSUE THREAD

**Issue by rpdelaney** [direnv/direnv#406]
## Issue description
I find that state data is stored in `$XDG_CONFIG_HOME/direnv/allow/`. 

Users who do not expect state data to be stored here may accidentally back it up and synchronize it across systems improperly. Configuration data are fine, but prefer storing state data somewhere in `$XDG_DATA_HOME` to conform to the freedesktop.org / XDG standard.

Reference: https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

```
$XDG_CONFIG_HOME defines the base directory relative to which user specific                                                                                                                                          
configuration files should be stored. If $XDG_CONFIG_HOME is either not set or                                                                                                                                       
empty, a default equal to $HOME/.config should be used.
```

```
$XDG_DATA_HOME defines the base directory relative to which user specific data                                                                                                                                       
files should be stored. If $XDG_DATA_HOME is either not set or empty, a default                                                                                                                                      
equal to $HOME/.local/share should be used.
```
### Steps to reproduce

`direnv allow .`

## Technical details

* `direnv version`: 2.17.0
* OS release: MacOS High Sierra (10.13.6)

**Comment by rpdelaney** [comment]
I don't see where the spec indicates that `$XDG_DATA_HOME` is meant for read-only data. Can you help me find that?

**Comment by zimbatm** [comment]
It doesn't say it explicitly unfortunately. Given the default path I assume that it inherits the same properties as /usr/share. See https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

**Comment by rpdelaney** [comment]
I don't think that assumption is well founded. Storing data the user is not able to change in the home directory would be the conflagration of the implicit conventions.

**Comment by zimbatm** [comment]
I don't know why it would be called "share" and have `$XDG_DATA_DIRS` also read in it, it's confusing. But looking at the content of that folder in my filesystem I can see that a bunch of application indeed use it as a sort of state folder.

**Comment by zimbatm** [comment]
https://github.com/direnv/direnv/pull/410 implements the change but breaks backward-compatibility, which I very much try to avoid doing. Any ideas?

**Comment by franckrasolo** [comment]
To preserve backward-compatibility, you could add a level of indirection by introducing an environment variable, say `DIRENV_DATA_DIR` that defaults to `$XDG_CONFIG_HOME/direnv` or `~/.config/direnv` if `XDG_CONFIG_HOME` is not set.

Users who want to adhere to the XDG spec would simply export `DIRENV_DATA_HOME` as `$XDG_DATA_HOME/direnv` as suggested by @rpdelaney above.

Alas, this is the compromise adopted by (far too) [many projects](https://wiki.archlinux.org/index.php/XDG_Base_Directory#Partial), leading to an [ever growing list](https://github.com/haskell/cabal/issues/680#issuecomment-113514145) of custom environment variables that users have to define.

I must say I'd rather have #410 as it stands together with [your commit message](https://github.com/direnv/direnv/pull/410/commits/96c92690fd5f6f192baaa2bee69bdcea7fb3efc9) added to both release notes and documentation.

**Comment by rpdelaney** [comment]
I'm not sure exactly how all this works, but could you do something like read the old location for state data (and accept it) but just stop writing it there?

**Comment by doronbehar** [comment]
Suggestion: How about making the binary in the next release move this state directory (if it exists) to `$XDG_DATA_HOME/direnv` by itslef? This behaviour should be stated in the manual of course, in addition to the fact that users who'd prefer the old behaviour will need to use `DIRENV_DATA_HOME`.

**Comment by SpotlightKid** [comment]
I got this after the latest direnv update:

```
$ direnv allow

Migrating the allow data to the new location

The allowed .envrc permissions used to be stored in the XDG_CONFIG_DIR. It's
better to keep that folder for user-editable configuration so the data is
being moved to XDG_DATA_HOME.

moving /home/chris/.config/direnv/allow to /home/chris/.local/share/direnv/allow
direnv: error rename /home/chris/.config/direnv/allow /home/chris/.local/share/direnv/allow: no such file or directory
```

Seems the migration process does not create a necessary directory `~/.local/share/direnv/`. I fixed this by running `mkdir ~/.local/share/direnv/` and then `direnv allow` again.

**Comment by rpdelaney** [comment]
Yeah, the destination is not being created, which requires manual intervention and troubleshooting of an obscure error message. We probably need to verify that the destination exists & is writable, then move the file(s), then verify that the move completed before we delete the data at the deprecated path.

**Comment by zimbatm** [comment]
this has been fixed in direnv 2.21.2

**Comment by tribals** [comment]
Hello everyone!

Personally I found freedesktop specs a little bit confusing. I'll cite right here to save time:

> `$XDG_DATA_HOME` defines the base directory relative to which user-specific data files should be stored. If `$XDG_DATA_HOME` is either not set or empty, a default equal to `$HOME/.local/share` should be used.
>
> `$XDG_CONFIG_HOME` defines the base directory relative to which user-specific configuration files should be stored. If `$XDG_CONFIG_HOME` is either not set or empty, a default equal to `$HOME/.config` should be used.
>
>`$XDG_STATE_HOME` defines the base directory relative to which user-specific state files should be stored. If `$XDG_STATE_HOME` is either not set or empty, a default equal to `$HOME/.local/state` should be used.
>
>The `$XDG_STATE_HOME` contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in `$XDG_DATA_HOME`. It may contain:
>
> - actions history (logs, history, recently used files, …)
>
> - current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

and

> `$XDG_CACHE_HOME` defines the base directory relative to which user-specific non-essential data files should be stored. If `$XDG_CACHE_HOME` is either not set or empty, a default equal to `$HOME/.cache` should be used.
>
> $XDG_RUNTIME_DIR defines the base directory relative to which user-specific non-essential runtime files and other file objects (such as sockets, named pipes, ...) should be stored.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

My best bet is that `allow/*` files should go to `$XDG_STATE_HOME` as it is essentially a _runtime state_ of an application that needs to be kept around reboots. And nobody backups state. Another option would be `$XDG_CACHE_HOME` - it is "meant to be" runtime data, and again nobody backups `$HOME/.cache`. But it definitely should not be `$XDG_DATA_HOME`. XDG spec is not clear about what data should go here. But if it defaults to `$HOME/.local/share`, and `$HOME/.local` is just another FHS level, then FHS is clear about what should go to `*/share`:

>        `/usr/share`
>              This directory contains subdirectories with specific application data, that *can be shared among different architectures of the same OS*.  Often one finds stuff here that used to live in `/usr/doc` or `/usr/lib` or `/usr/man`.

and

>        `/usr/local/share`
>              Local application data that can be shared among different architectures of the same OS.

`man 7 hier`

Definitely, `allow/*` files neither want to nor should be shared.
